From 9284130f3494079cdfbb9aed9b4de6f433b3c42f Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:32:20 -0400
Subject: [iommu] vt-d: fix a race window in allocating domain ID for virtual machines

Message-id: <20140905013220.30867.24898.stgit@gir.stowe>
Patchwork-id: 91276
O-Subject: [RHEL7.1 PATCH 12/28] iommu/vt-d: fix a race window in allocating domain ID for virtual machines
Bugzilla: 1129808
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129808
Upstream Status: 18d99165d3ebe5e365de57bcc673901d754c7142

commit 18d99165d3ebe5e365de57bcc673901d754c7142
Author: Jiang Liu <jiang.liu@linux.intel.com>
Date:   Mon Jan 6 14:18:10 2014 +0800

    iommu/vt-d: fix a race window in allocating domain ID for virtual machines

    Function intel_iommu_domain_init() may be concurrently called by upper
    layer without serialization, so use atomic_t to protect domain id
    allocation.

    Signed-off-by: Jiang Liu <jiang.liu@linux.intel.com>
    Cc: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/intel-iommu.c b/drivers/iommu/intel-iommu.c
index 04fbb88..1f56039 100644
--- a/drivers/iommu/intel-iommu.c
+++ b/drivers/iommu/intel-iommu.c
@@ -3877,7 +3877,7 @@ static void vm_domain_remove_all_dev_info(struct dmar_domain *domain)
 }
 
 /* domain id for virtual machine, it won't be set in context */
-static unsigned long vm_domid;
+static atomic_t vm_domid = ATOMIC_INIT(0);
 
 static struct dmar_domain *iommu_alloc_vm_domain(void)
 {
@@ -3887,7 +3887,7 @@ static struct dmar_domain *iommu_alloc_vm_domain(void)
  if (!domain)
   return NULL;
 
- domain->id = vm_domid++;
+ domain->id = atomic_inc_return(&vm_domid);
  domain->nid = -1;
  memset(domain->iommu_bmp, 0, sizeof(domain->iommu_bmp));
  domain->flags = DOMAIN_FLAG_VIRTUAL_MACHINE;
-- 
1.7.1